tools/autograd/derivatives.yaml

```diff
@@ -1274,7 +1274,7 @@
 - name: remainder.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad
   other: -grad * self.div(other, /*rounding_mode=*/"floor")
-  result: self_t - other_t * self_p.div(other_p, /*rounding_mode=*/"floor")
```

Comment:
I'd suggest not adding `remainder_jvp` and instead modifying this formula to:

`self_t - (other_t * self_p).div(other_p, /*rounding_mode=*/"floor")`

Reply:
yep yep, that was just for testing. to make sure i can hit the ZT path when measuring perf during profiling

Reply:
fun fact: the change you're proposing here is actually slower than the non-zt version and also slower than the version with an if. here you can see two tests using the updated profiling code with the dual and with an undefined tensor. you can spot where ZT is used by looking at the stack trace.

this is likely due to div and mul not short-circuiting with a zt argument.

```
zt:

---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                 aten::_make_dual        15.59%      53.000us        22.65%      77.000us      77.000us             1  
                      aten::alias         4.71%      16.000us         4.71%      16.000us       5.333us             3  
               aten::is_same_size         0.88%       3.000us         0.88%       3.000us       1.500us             2  
    aten::_has_same_storage_numel         1.76%       6.000us         1.76%       6.000us       3.000us             2  
                  aten::remainder        25.59%      87.000us        77.35%     263.000us     263.000us             1  
                 aten::_fw_primal         1.47%       5.000us         1.47%       5.000us       2.500us             2  
       aten::_efficientzerotensor         0.88%       3.000us         0.88%       3.000us       1.500us             2  
                        aten::mul         7.35%      25.000us        35.88%     122.000us     122.000us             1  
                         aten::to         6.47%      22.000us        28.24%      96.000us      48.000us             2  
                   aten::_to_copy        10.29%      35.000us        21.76%      74.000us      37.000us             2  
              aten::empty_strided         2.94%      10.000us         2.94%      10.000us      10.000us             1  
                      aten::copy_         1.76%       6.000us         1.76%       6.000us       3.000us             2  
                      aten::zeros         3.53%      12.000us         5.00%      17.000us       8.500us             2  
                      aten::empty         1.18%       4.000us         1.18%       4.000us       1.333us             3  
                      aten::zero_         0.59%       2.000us         0.59%       2.000us       1.000us             2  
                     aten::expand         2.06%       7.000us         2.35%       8.000us       4.000us             2  
                 aten::as_strided         0.29%       1.000us         0.29%       1.000us       0.500us             2  
                        aten::div         7.65%      26.000us         8.53%      29.000us      29.000us             1  
                        aten::sub         5.00%      17.000us         5.00%      17.000us      17.000us             1  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 340.000us

non-zt:

---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                 aten::_make_dual        27.35%      61.000us        39.01%      87.000us      43.500us             2  
                      aten::alias         7.17%      16.000us         7.17%      16.000us       4.000us             4  
               aten::is_same_size         1.35%       3.000us         1.35%       3.000us       1.000us             3  
    aten::_has_same_storage_numel         3.59%       8.000us         3.59%       8.000us       2.667us             3  
                  aten::remainder        37.67%      84.000us        60.99%     136.000us     136.000us             1  
                 aten::_fw_primal         2.24%       5.000us         2.24%       5.000us       2.500us             2  
                        aten::mul         6.73%      15.000us         6.73%      15.000us      15.000us             1  
                        aten::div         8.07%      18.000us         8.07%      18.000us      18.000us             1  
                        aten::sub         5.83%      13.000us         5.83%      13.000us      13.000us             1  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 223.000us
```

Reply:
so i'll go with an inline version that uses an if

Reply:
with an if:

```
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                 aten::_make_dual        34.78%      56.000us        49.07%      79.000us      79.000us             1  
                      aten::alias         7.45%      12.000us         7.45%      12.000us       4.000us             3  
               aten::is_same_size         1.86%       3.000us         1.86%       3.000us       1.500us             2  
    aten::_has_same_storage_numel         4.97%       8.000us         4.97%       8.000us       4.000us             2  
                  aten::remainder        46.58%      75.000us        50.93%      82.000us      82.000us             1  
                 aten::_fw_primal         3.11%       5.000us         3.11%       5.000us       2.500us             2  
       aten::_efficientzerotensor         1.24%       2.000us         1.24%       2.000us       2.000us             1  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 161.000us
```

Reply:
also note that the perf speed up is relatively minor here (unless it accumulates over many calls like that) and the majority of it comes from not calling make_dual to pass an undefined tensor

Reply:
huh I am not sure why `mul_zerotensor` and `div_zerotensor` don't get called when in `other_t` is a ZT in this case:
`self_t - (other_t * self_p).div(other_p, /*rounding_mode=*/"floor")`

I would expect it to call into those two kernels because the multiplication inside the bracket would get evaluated first and return a ZT.

cc. @albanD any idea what's going on here?

Reply:
Could it be because you're doing 2 unnecessary dispatches vs none with the ternary operator?

Reply:
also, just to confirm. this with an updated benchmark posted here:

https://github.com/pytorch/pytorch/pull/76479#issuecomment-1116663454

with the change suggested in this thread:

```
self_t - (other_t * self_p).div(other_p, /*rounding_mode=*/"floor")
```

no zt:
```
tensor([ 0.1184,  0.0486,  0.3453,  0.2018,  0.2039,  1.4944, -0.5317, -1.1885,
        -0.3465,  1.7781])
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                  aten::remainder        67.35%      99.000us       100.00%     147.000us     147.000us             1  
                 aten::_fw_primal         5.44%       8.000us         6.80%      10.000us       5.000us             2  
                      aten::alias         1.36%       2.000us         1.36%       2.000us       1.000us             2  
                        aten::mul         9.52%      14.000us         9.52%      14.000us      14.000us             1  
                        aten::div         7.48%      11.000us         7.48%      11.000us      11.000us             1  
                        aten::sub         7.48%      11.000us         7.48%      11.000us      11.000us             1  
               aten::is_same_size         0.68%       1.000us         0.68%       1.000us       1.000us             1  
    aten::_has_same_storage_numel         0.68%       1.000us         0.68%       1.000us       1.000us             1  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 147.000us
```

zt:
```
tensor([-0.4681, -0.4405,  0.7630, -0.0720, -1.0708, -0.0447,  0.2598, -0.0314,
         0.2920, -0.3969])
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                             Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                  aten::remainder        37.20%     109.000us       100.00%     293.000us     293.000us             1  
                 aten::_fw_primal         3.41%      10.000us         5.12%      15.000us       7.500us             2  
                      aten::alias         1.71%       5.000us         1.71%       5.000us       2.500us             2  
       aten::_efficientzerotensor         1.71%       5.000us         1.71%       5.000us       2.500us             2  
                        aten::mul         8.53%      25.000us        42.66%     125.000us     125.000us             1  
                         aten::to         8.87%      26.000us        34.13%     100.000us      50.000us             2  
                   aten::_to_copy        11.26%      33.000us        25.26%      74.000us      37.000us             2  
              aten::empty_strided         3.07%       9.000us         3.07%       9.000us       9.000us             1  
                      aten::copy_         1.71%       5.000us         1.71%       5.000us       2.500us             2  
                      aten::zeros         5.12%      15.000us         7.17%      21.000us      10.500us             2  
                      aten::empty         1.02%       3.000us         1.02%       3.000us       1.000us             3  
                      aten::zero_         1.37%       4.000us         1.37%       4.000us       2.000us             2  
                     aten::expand         2.39%       7.000us         2.73%       8.000us       4.000us             2  
                 aten::as_strided         0.34%       1.000us         0.34%       1.000us       0.500us             2  
                        aten::div         7.51%      22.000us         8.53%      25.000us      25.000us             1  
                        aten::sub         4.10%      12.000us         4.10%      12.000us      12.000us             1  
               aten::is_same_size         0.34%       1.000us         0.34%       1.000us       1.000us             1  
    aten::_has_same_storage_numel         0.34%       1.000us         0.34%       1.000us       1.000us             1  
---------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 293.000us
```

i mean this is not surprising given that it was possible to see things not being optimized out in the previous call stack, but i just wanted to profile it again.

so there's definitely a problem with mul/div not short-circuiting on zt as they should. so i'll go with the inline version that uses an if.

Reply:
also, this is a similar benchmark with cuda (hope i did this right) for

```
self_t - (other_t * self_p).div(other_p, /*rounding_mode=*/"floor")
```

(on the same commit too)

looks like there's mostly no difference on cuda. the number of calls is different and one of the calls is not vectorized, but that's it.

no zt:
```
tensor([ 0.3272,  0.3525,  0.0707, -0.0189, -0.0435,  0.6581,  0.0894,  0.1752,
        -0.0163, -0.6054], device='cuda:0')
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                       cudaLaunchKernel        84.62%      55.000us        84.62%      55.000us      13.750us       0.000us         0.00%       0.000us       0.000us             4  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.000us        30.00%       3.000us       3.000us             1  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us        20.00%       2.000us       2.000us             1  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.000us        30.00%       3.000us       3.000us             1  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us        20.00%       2.000us       2.000us             1  
                                  cudaDeviceSynchronize        15.38%      10.000us        15.38%      10.000us      10.000us       0.000us         0.00%       0.000us       0.000us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 65.000us
Self CUDA time total: 10.000us
```

zt:
```
(pytorch-dev) [nkaretnikov@qgpu3:~/pytorch]% python ../bench_zt_remainder4_2.py                                                                                                                                                                      -- INSERT --
tensor([-0.0067,  1.2473,  0.0111, -0.2884,  0.0951,  0.1896,  0.7708,  0.3875,
         0.5270,  0.5987], device='cuda:0')
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                       cudaLaunchKernel        87.14%      61.000us        87.14%      61.000us      12.200us       0.000us         0.00%       0.000us       0.000us             5  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       3.000us        23.08%       3.000us       3.000us             1  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       4.000us        30.77%       4.000us       2.000us             2  
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us       4.000us        30.77%       4.000us       4.000us             1  
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us        15.38%       2.000us       2.000us             1  
                                  cudaDeviceSynchronize        12.86%       9.000us        12.86%       9.000us       9.000us       0.000us         0.00%       0.000us       0.000us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 70.000us
Self CUDA time total: 13.000us
```

benchmark:

```py
import torch
import torch.autograd.forward_ad as fwAD

self_p = torch.randn(10, device='cuda')
self_t = torch.randn(10, device='cuda')
other_p = torch.randn(10, device='cuda')
other_t = torch.randn(10, device='cuda')

with fwAD.dual_level():
    dual_inp_x = fwAD.make_dual(self_p, self_t)
    dual_inp_y = other_p
    # dual_inp_y = fwAD.make_dual(other_p, other_t)
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CUDA
        ]
    ) as p:
        dual_output = torch.remainder(dual_inp_x, dual_inp_y)

print(dual_output)
print(p.key_averages())

```